### PR TITLE
New version of IIDM current limits

### PR DESCRIPTION
**Pull request by geofjamg**

### CODE REVIEW

File: iidm-network-impl/src/main/java/eu/itesla_project/iidm/network/impl/AbstractTwoTerminalsConnectable.java [hunk 1]
```diff
@@ -92,27 +92,129 @@ public CurrentLimitsAdder newCurrentLimits2() {
...
-            float i1 = t.getI();
-            float previousLimit = limits.getPermanentLimit();
-            for (CurrentLimits.TemporaryLimit tl : limits.getTemporaryLimits()) {
-                if (i1 >= previousLimit && i1 < tl.getLimit()) {
-                    return tl.getAcceptableDuration();
-                }
-                previousLimit = tl.getLimit();
+    public int getOverloadDuration() {
+        TwoTerminalsConnectable.Overload o1 = checkTemporaryLimits1();
+        TwoTerminalsConnectable.Overload o2 = checkTemporaryLimits1();
```
Comment (mathbagu) [inline iidm-network-impl/src/main/java/eu/itesla_project/iidm/network/impl/AbstractTwoTerminalsConnectable.java]: Copy/paste error : checkTemporaryLimits2?

File: iidm-network-impl/src/main/java/eu/itesla_project/iidm/network/impl/AbstractTwoTerminalsConnectable.java [hunk 2]
```diff
@@ -92,27 +92,129 @@ public CurrentLimitsAdder newCurrentLimits2() {
...
+    public boolean checkPermanentLimit2(float limitReduction) {
+        return checkPermanentLimit(getTerminal2(), getCurrentLimits2(), limitReduction);
+    }
+
+    public boolean checkPermanentLimit2() {
+        return checkPermanentLimit2(1f);
+    }
+
+    public static boolean checkPermanentLimit(Terminal terminal, CurrentLimits limits, float limitReduction) {
+        float i = terminal.getI();
```
Comment (mathbagu) [inline iidm-network-impl/src/main/java/eu/itesla_project/iidm/network/impl/AbstractTwoTerminalsConnectable.java]: is Objects.requireNonNull(terminal) needed?

File: iidm-xml-converter/src/main/java/eu/itesla_project/iidm/xml/ConnectableXml.java [hunk 3]
```diff
@@ -128,11 +128,15 @@ public static void readCurrentLimits(Integer index, Supplier<CurrentLimitsAdder>
         adder.setPermanentLimit(permanentLimit);
         XmlUtil.readUntilEndElement("currentLimits" + indexToString(index), reader, () -> {
             if ("temporaryLimit".equals(reader.getLocalName())) {
-                int acceptableDuration = Integer.valueOf(reader.getAttributeValue(null, "acceptableDuration"));
-                float limit = Float.valueOf(XmlUtil.readUntilEndElement("temporaryLimit", reader, null));
+                String name = reader.getAttributeValue(null, "name");
+                Integer acceptableDuration = XmlUtil.readOptionalIntegerAttribute(reader, "acceptableDuration");
```
Comment (mathbagu) [inline iidm-xml-converter/src/main/java/eu/itesla_project/iidm/xml/ConnectableXml.java]: Could be replaced by int acceptableDuration = XmlUtil.readOptionalIntegerAttribute(reader, "acceptableDuration", Integer.MAX_VALUE); ?

File: iidm-xml-converter/src/main/java/eu/itesla_project/iidm/xml/ConnectableXml.java [hunk 4]
```diff
@@ -128,11 +128,15 @@ public static void readCurrentLimits(Integer index, Supplier<CurrentLimitsAdder>
...
+                float value = XmlUtil.readOptionalFloatAttribute(reader, "value");
```
Comment (mathbagu) [inline iidm-xml-converter/src/main/java/eu/itesla_project/iidm/xml/ConnectableXml.java]: Could be replaced by float value = XmlUtil.readOptionalFloatAttribute(reader, "value", Float.MAX_VALUE); ?
Maybe not due to "Float.isNaN(value)" check below.

File: security-analysis/src/main/java/eu/itesla_project/security/LimitViolation.java [hunk 5]
```diff
@@ -30,18 +32,19 @@
 
     private final float baseVoltage;
 
-    public LimitViolation(Identifiable subject, LimitViolationType limitType, float limit, float limitReduction, float value, Country country, float baseVoltage) {
+    public LimitViolation(Identifiable subject, LimitViolationType limitType, float limit, String limitName, float limitReduction, float value, Country country, float baseVoltage) {
         this.subject = Objects.requireNonNull(subject);
         this.limitType = Objects.requireNonNull(limitType);
         this.limit = limit;
+        this.limitName = limitName;
```
Comment (mathbagu) [inline security-analysis/src/main/java/eu/itesla_project/security/LimitViolation.java]: Add a test to check if limitName is null or not?

File: security-analysis/src/main/java/eu/itesla_project/security/Security.java [hunk 6]
```diff
@@ -28,75 +28,93 @@ public Security() {
...
-                        break;
-                    }
-                }
-                break;
-
-            default:
-                throw new AssertionError();
-        }
-        return limit;
+    private static Country getCountry(TwoTerminalsConnectable branch, Terminal terminal) {
```
Comment (mathbagu) [inline security-analysis/src/main/java/eu/itesla_project/security/Security.java]: Add a test to check if branch or terminal are null?

File: security-analysis/src/main/java/eu/itesla_project/security/Security.java [hunk 7]
```diff
@@ -28,75 +28,93 @@ public Security() {
...
-                        Country country = terminal == branch.getTerminal1() ? branch.getTerminal1().getVoltageLevel().getSubstation().getCountry()
-                                : branch.getTerminal2().getVoltageLevel().getSubstation().getCountry();
-                        float baseVoltage = Math.max(branch.getTerminal1().getVoltageLevel().getNominalV(),
-                                                     branch.getTerminal2().getVoltageLevel().getNominalV());
-                        violations.add(new LimitViolation(branch, LimitViolationType.CURRENT, limit, limitReduction, i, country, baseVoltage));
-                    }
-                }
-            }
-        }
+    private static float getBaseVoltage(TwoTerminalsConnectable branch) {
```
Comment (mathbagu) [inline security-analysis/src/main/java/eu/itesla_project/security/Security.java]: Add a test to check if branch is null?

File: security-analysis/src/main/java/eu/itesla_project/security/Security.java [hunk 8]
```diff
@@ -28,75 +28,93 @@ public Security() {
...
+        return Math.max(branch.getTerminal1().getVoltageLevel().getNominalV(),
+                        branch.getTerminal2().getVoltageLevel().getNominalV());
     }
 
     public static void checkCurrentLimits(Iterable<? extends TwoTerminalsConnectable> branches, CurrentLimitType currentLimitType,
-                                          int maxAcceptableDuration, float limitReduction, List<LimitViolation> violations) {
+                                          float limitReduction, List<LimitViolation> violations) {
```
Comment (mathbagu) [inline security-analysis/src/main/java/eu/itesla_project/security/Security.java]: Add test to check if parameters are null or not?

File: iidm-network-impl/src/main/java/eu/itesla_project/iidm/network/impl/AbstractTwoTerminalsConnectable.java (hunk 1 above)
Comment (geofjamg) [inline iidm-network-impl/src/main/java/eu/itesla_project/iidm/network/impl/AbstractTwoTerminalsConnectable.java]: Yes

File: iidm-network-impl/src/main/java/eu/itesla_project/iidm/network/impl/AbstractTwoTerminalsConnectable.java (hunk 2 above)
Comment (geofjamg) [inline iidm-network-impl/src/main/java/eu/itesla_project/iidm/network/impl/AbstractTwoTerminalsConnectable.java]: OK, also changed method to private

File: iidm-xml-converter/src/main/java/eu/itesla_project/iidm/xml/ConnectableXml.java (hunk 3 above)
Comment (geofjamg) [inline iidm-xml-converter/src/main/java/eu/itesla_project/iidm/xml/ConnectableXml.java]: OK

File: iidm-xml-converter/src/main/java/eu/itesla_project/iidm/xml/ConnectableXml.java (hunk 4 above)
Comment (geofjamg) [inline iidm-xml-converter/src/main/java/eu/itesla_project/iidm/xml/ConnectableXml.java]: OK, need to add XmlUtil.readOptionalFloatAttribute supporting default value

File: security-analysis/src/main/java/eu/itesla_project/security/LimitViolation.java (hunk 5 above)
Comment (geofjamg) [inline security-analysis/src/main/java/eu/itesla_project/security/LimitViolation.java]: No, limitName is optional.

File: security-analysis/src/main/java/eu/itesla_project/security/Security.java (hunk 6 above)
Comment (geofjamg) [inline security-analysis/src/main/java/eu/itesla_project/security/Security.java]: Not really necessary because private method

File: security-analysis/src/main/java/eu/itesla_project/security/Security.java (hunk 7 above)
Comment (geofjamg) [inline security-analysis/src/main/java/eu/itesla_project/security/Security.java]: Not really necessary because private method

File: security-analysis/src/main/java/eu/itesla_project/security/Security.java (hunk 8 above)
Comment (geofjamg) [inline security-analysis/src/main/java/eu/itesla_project/security/Security.java]: I changed this method to private and added null checks to public methods checkLimits and printLimitsViolations
